### DISCUSSION TITLE
Time synchronization in TDoA

### DISCUSSION THREAD

**Discussion by danteppc** [bitcraze/discussions#143] [Q&A] [answered]
Hi,

According to the documents, in TDoA2 the anchors' clocks are [synchronized based on 'anchor 0'](https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/protocols/tdoa2_protocol/) and in the TDoA3 mode the [anchors' clocks are not synchronized](https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/protocols/tdoa3_protocol/).

My question is will the tag's (Crazyflie) clock be synchronized with 'anchor 0' as well in TDoA2 or maybe TDoA3? if not, is there an option to enable time synchronization between tags and anchors? 

Thank you

**Comment by knmcguire** [comment]
I think that @krichardsson is the expert here, but since he is still on a holiday I'll try to help out.

In both TDOA2 and 3, the loco nodes also range with each other in order for everything to work. 

- In TDOA2, the ranging is done in a round robin fashion, so first 0 - 1 range, than 1 - 2, than 2- 3 and so on. 
- In TDOA3 the ranging is done at different frequencies with a small random seed around the ranging frequency, so semi randomly 0-1 ranges, than 2-4 ranges. than 6-8 ranges, and so on.  

TDOA2 node ranging is quite fast, but as soon one node is disconnect this fails and it's is not flexible for adding more nodes. TDOA3 suffers from more packet loss due to collisions, but it's easier to add more nodes and more robust if one of the nodes fail. 

It's a bit of a nutshell explanation but I hope it is clear.  @krichardsson wrote a blogpost of this quite recently which is interesting to read as well: https://www.bitcraze.io/2022/06/the-loco-positioning-system/

**Reply by danteppc** [reply]
Thank you for your response,

So in TDoA2 the clocks between the anchors are kept in sync by the ranging signals, but my question is: what about Crazyflie itself?

Does the clock in Crazyflie need to be synchronized in order for TDoA work?

**Reply by danteppc** [reply]
Ok so the answer is probably "no" since TDoA only works with the difference not absolute values.

My further question is, can we enable time synchronization of tags on top of the broadcasting ranging signals? similar to GPS?

Thanks

**Comment by knmcguire** [comment]
> My further question is, can we enable time synchronization of tags on top of the broadcasting ranging signals? similar to GPS?

Let me start a new thread here. 

Maybe this is possible with [two way ranging mode](https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/protocols/twr-protocol/)? 

I unfortunately do not know enough about this, but kristoffer is back in a week so I'll ping him them to answer this

**Comment by krichardsson** [comment, marked as answer]
Hi, I'm back from my holidays, let's clear this up. 

There are a few clocks involved here, one clock in each DWM module (Loco decks and anchors) and one for each CPU. The sentence "clocks are synchronized based on 'anchor 0'" is pretty ambiguous :-) 

What we mean in the documentation is that the transmission time of each anchor is time slotted in a round robin fashion as @knmcguire  mentioned, and based on when anchor 0 transmits. The scheme is pretty simple:
* The time slots are 2 ms (I think)
* The system supports 8 anchors
* Anchor 0 transmits every 8 * 2 ms
* The other anchors listen to when anchor 0 transmits and schedules their own transmission time to be ID * 2 ms later (where ID is the anchor id, 1 to 7)

The high frequency DWM clocks are not fully synchronized, but their [relative difference in clock rate is derived from the traffic](https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/functional-areas/tdoa3_implementation/#clock-correction) in the same way as is done in TDoA3. Most of the code in the Crazyflie firmware for TDoA2 and 3 is actually shared, only the part that is handling the protocol is different.

> is there an option to enable time synchronization between tags and anchors?

Nope, not at this time as it is not needed for estimating the position. It is maybe possible to implement some sort of PLL in the Crazyflie that tracks the remote clock (both phase and rate) but it is pretty tricky.
